### PR TITLE
feat: add update margin mode transaction

### LINKED ISSUE #9985: [Feature]: Add a transaction for update margin mode
# Feature Overview

**In order to** be able to switch between cross an isolated margin
**We will** add a new transaction type
**So that** a party can margin mode can be changed. 

## Specs
- [[Update margin mode ](https://github.com/vegaprotocol/specs/blob/palazzo/protocol/0019-MCAL-margin_calculator.md#isolated-margin-mode)]

### PR DESCRIPTION
**Pull request by ze97286**
Closes #9985 

This is the first PR in a series for adding support for isolated margin. In this PR the new transaction is added and validated up to the point that it can be sent to the market for final checks, but still without support in the market. The following PRs will add support for the margin calculation and being able to switch the mode.

### CODE REVIEW

**Review by EVODelavega — DISMISSED** [review]

**Review by EVODelavega — DISMISSED** [review]

File: protos/sources/vega/commands/v1/transaction.proto [hunk 1]
```diff
@@ -61,6 +61,8 @@ message InputData {
     UpdateReferralSet update_referral_set = 1019;
     // Command to apply a referral code.
     ApplyReferralCode apply_referral_code = 1020;
+    // Command to update the margin mode of a party in a market.
+    UpdateMarginMode update_margine_mode = 1021;
```
Comment (jeremyletang) [inline protos/sources/vega/commands/v1/transaction.proto:65]:
```suggestion
    UpdateMarginMode update_margin_mode = 1021;
```

File: commands/transaction.go [hunk 2]
```diff
@@ -250,6 +250,8 @@ func CheckInputData(rawInputData []byte) (*commandspb.InputData, Errors) {
 			errs.Merge(checkUpdateReferralSet(cmd.UpdateReferralSet))
 		case *commandspb.InputData_ApplyReferralCode:
 			errs.Merge(checkApplyReferralCode(cmd.ApplyReferralCode))
+		case *commandspb.InputData_UpdateMargineMode:
+			errs.Merge(checkUpdateMarginMode(cmd.UpdateMargineMode))
```
Comment (candida-d) [inline commands/transaction.go:254]:
```suggestion
		case *commandspb.InputData_UpdateMarginMode:
			errs.Merge(checkUpdateMarginMode(cmd.UpdateMarginMode))
```

File: core/processor/tx.go [hunk 3]
```diff
@@ -416,6 +418,12 @@ func (t Tx) Unmarshal(i interface{}) error {
 			return errors.New("failed to unmarshall to JoinTeam")
 		}
 		*underlyingCmd = *cmd.ApplyReferralCode
+	case *commandspb.InputData_UpdateMargineMode:
```
Comment (candida-d) [inline core/processor/tx.go:421]:
```suggestion
	case *commandspb.InputData_UpdateMarginMode:
```

File: core/processor/tx.go [hunk 4]
```diff
@@ -229,6 +229,8 @@ func (t Tx) GetCmd() interface{} {
 		return cmd.UpdateReferralSet
 	case *commandspb.InputData_ApplyReferralCode:
 		return cmd.ApplyReferralCode
+	case *commandspb.InputData_UpdateMargineMode:
```
Comment (candida-d) [inline core/processor/tx.go:232]:
```suggestion
	case *commandspb.InputData_UpdateMarginMode:
```

File: core/processor/tx.go [hunk 5]
```diff
@@ -229,6 +229,8 @@ func (t Tx) GetCmd() interface{} {
...
+		return cmd.UpdateMargineMode
```
Comment (candida-d) [inline core/processor/tx.go:233]:
```suggestion
		return cmd.UpdateMarginMode
```

File: core/processor/tx.go [hunk 6]
```diff
@@ -416,6 +418,12 @@ func (t Tx) Unmarshal(i interface{}) error {
 			return errors.New("failed to unmarshall to JoinTeam")
 		}
 		*underlyingCmd = *cmd.ApplyReferralCode
+	case *commandspb.InputData_UpdateMargineMode:
+		underlyingCmd, ok := i.(*commandspb.UpdateMarginMode)
+		if !ok {
+			return errors.New("failed to unmarshall to UpdateMarginMode")
+		}
+		*underlyingCmd = *cmd.UpdateMargineMode
```
Comment (candida-d) [inline core/processor/tx.go:426]:
```suggestion
		*underlyingCmd = *cmd.UpdateMarginMode
```

File: wallet/commands/commands.go [hunk 7]
```diff
@@ -228,6 +230,10 @@ func WrapRequestCommandIntoInputData(data *commandspb.InputData, req *walletpb.S
 		data.Command = &commandspb.InputData_ApplyReferralCode{
 			ApplyReferralCode: req.GetApplyReferralCode(),
 		}
+	case *walletpb.SubmitTransactionRequest_UpdateMarginMode:
+		data.Command = &commandspb.InputData_UpdateMargineMode{
+			UpdateMargineMode: req.GetUpdateMarginMode(),
```
Comment (candida-d) [inline wallet/commands/commands.go:235]:
```suggestion
		data.Command = &commandspb.InputData_UpdateMarginMode{
			UpdateMarginMode: req.GetUpdateMarginMode(),
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 8]
```diff
@@ -106,6 +106,17 @@ message IcebergOpts {
...
 
+message UpdateMarginMode {
+  enum Mode {
+    MODE_CROSS_UNSPECIFIED = 0;
+    MODE_CROSS_MARGIN = 1;
+    MODE_ISOLATED_MARGIN = 2;
+  }
+  string market_id = 1;
+  Mode mode = 2;
+  string margin_factor = 3;
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:117]: Docs please!

File: protos/sources/vega/vega.proto [hunk 9]
```diff
@@ -1774,3 +1777,9 @@ message ActivityStreakBenefitTier {
   // Vesting bonus applicable to this tier.
   string vesting_multiplier = 3;
 }
+
+enum MarginMode {
+  MARGIN_MODE_UNSPECIFIED = 0;
+  MARGIN_MODE_CROSS_MARGIN = 1;
+  MARGIN_MODE_ISOLATED_MARGIN = 2;
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1784]: here too please

File: protos/vega/commands/v1/transaction.pb.go [hunk 10]
```diff
@@ -313,6 +314,13 @@ func (x *InputData) GetApplyReferralCode() *ApplyReferralCode {
 	return nil
 }
 
+func (x *InputData) GetUpdateMargineMode() *UpdateMarginMode {
+	if x, ok := x.GetCommand().(*InputData_UpdateMargineMode); ok {
+		return x.UpdateMargineMode
```
Comment (candida-d) [inline protos/vega/commands/v1/transaction.pb.go:319]:
```suggestion
func (x *InputData) GetUpdateMarginMode() *UpdateMarginMode {
	if x, ok := x.GetCommand().(*InputData_UpdateMarginMode); ok {
		return x.UpdateMarginMode
```

File: protos/vega/commands/v1/transaction.pb.go [hunk 11]
```diff
@@ -487,6 +495,11 @@ type InputData_ApplyReferralCode struct {
 	ApplyReferralCode *ApplyReferralCode `protobuf:"bytes,1020,opt,name=apply_referral_code,json=applyReferralCode,proto3,oneof"`
 }
 
+type InputData_UpdateMargineMode struct {
+	// Command to update the margin mode of a party in a market.
+	UpdateMargineMode *UpdateMarginMode `protobuf:"bytes,1021,opt,name=update_margine_mode,json=updateMargineMode,proto3,oneof"`
```
Comment (candida-d) [inline protos/vega/commands/v1/transaction.pb.go:500]:
```suggestion
type InputData_UpdateMarginMode struct {
	// Command to update the margin mode of a party in a market.
	UpdateMarginMode *UpdateMarginMode `protobuf:"bytes,1021,opt,name=update_margin_mode,json=updateMarginMode,proto3,oneof"`
```

File: protos/vega/commands/v1/transaction.pb.go [hunk 12]
```diff
@@ -1074,24 +1096,25 @@ var file_vega_commands_v1_transaction_proto_depIdxs = []int32{
...
-	33, // 29: vega.commands.v1.InputData.oracle_data_submission:type_name -> vega.commands.v1.OracleDataSubmission
-	34, // 30: vega.commands.v1.Transaction.signature:type_name -> vega.commands.v1.Signature
-	0,  // 31: vega.commands.v1.Transaction.version:type_name -> vega.commands.v1.TxVersion
-	3,  // 32: vega.commands.v1.Transaction.pow:type_name -> vega.commands.v1.ProofOfWork
-	33, // [33:33] is the sub-list for method output_type
-	33, // [33:33] is the sub-list for method input_type
-	33, // [33:33] is the sub-list for extension type_name
-	33, // [33:33] is the sub-list for extension extendee
-	0,  // [0:33] is the sub-list for field type_name
+	24, // 20: vega.commands.v1.InputData.update_margine_mode:type_name -> vega.commands.v1.UpdateMarginMode
```
Comment (candida-d) [inline protos/vega/commands/v1/transaction.pb.go:1099]:
```suggestion
	24, // 20: vega.commands.v1.InputData.update_margin_mode:type_name -> vega.commands.v1.UpdateMarginMode
```

File: protos/vega/commands/v1/transaction.pb.go [hunk 13]
```diff
@@ -1162,6 +1185,7 @@ func file_vega_commands_v1_transaction_proto_init() {
 		(*InputData_CreateReferralSet)(nil),
 		(*InputData_UpdateReferralSet)(nil),
 		(*InputData_ApplyReferralCode)(nil),
+		(*InputData_UpdateMargineMode)(nil),
```
Comment (candida-d) [inline protos/vega/commands/v1/transaction.pb.go:1188]:
```suggestion
		(*InputData_UpdateMarginMode)(nil),
```

File: protos/vega/commands/v1/transaction.pb.go [hunk 14]
```diff
@@ -577,6 +590,8 @@ func (*InputData_UpdateReferralSet) isInputData_Command() {}
 
 func (*InputData_ApplyReferralCode) isInputData_Command() {}
 
+func (*InputData_UpdateMargineMode) isInputData_Command() {}
```
Comment (candida-d) [inline protos/vega/commands/v1/transaction.pb.go:593]:
```suggestion
func (*InputData_UpdateMarginMode) isInputData_Command() {}
```

File: protos/vega/commands/v1/transaction.pb.go [hunk 15]
```diff
@@ -107,6 +107,7 @@ type InputData struct {
 	//	*InputData_CreateReferralSet
 	//	*InputData_UpdateReferralSet
 	//	*InputData_ApplyReferralCode
+	//	*InputData_UpdateMargineMode
```
Comment (candida-d) [inline protos/vega/commands/v1/transaction.pb.go:110]:
```suggestion
	//	*InputData_UpdateMarginMode
```

File: protos/sources/vega/commands/v1/commands.proto (hunk 8 above)
Comment (wwestgarth) [inline protos/sources/vega/commands/v1/commands.proto:117]: `maring_factor` should probably be `optional` if it only applies to one mode.

**Review by jeremyletang — DISMISSED** [review]

File: protos/sources/vega/commands/v1/commands.proto [hunk 16]
```diff
@@ -106,6 +106,23 @@ message IcebergOpts {
...
+    // Cross margin mode - margin is dynamically acquired and released as a position is marked to market
+    MODE_CROSS_MARGIN = 1;
+    // Isolated margin mode - margin for any newly opened position volume is transferred to the margin account when the trade is executed
+    MODE_ISOLATED_MARGIN = 2;
+  }
+  // Market to change margin mode for.
+  string market_id = 1;
+  // Margin mode to use.
+  Mode mode = 2;
+  // Margin factor to use for margin in isolated mode.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:122]:
```suggestion
  // Margin factor to use for margin in isolated mode. It is a multiplier that defines how much margin needs to be set aside
```

**Review by candida-d — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**